rust/src/fedora_integration.rs

```diff
@@ -0,0 +1,151 @@
+
+mod bodhi {
+    use super::*;
+
+    #[derive(Deserialize)]
+    pub(crate) struct BodhiKojiBuild {
+        pub(crate) nvr: String,
+        /// Not currently included in koji URLs, so we ignore it
+        #[allow(dead_code)]
+        pub(crate) epoch: u64,
+    }
+
+    #[derive(Deserialize)]
+    pub(crate) struct BodhiUpdate {
+        pub(crate) builds: Vec<BodhiKojiBuild>,
+    }
+
+    #[derive(Deserialize)]
+    struct BodhiUpdateResponse {
+        update: BodhiUpdate,
+    }
+
+    pub(crate) fn canonicalize_update_id(id: &str) -> Cow<str> {
+        let id = match id.strip_prefix(BODHI_URL_PREFIX) {
+            Some(s) => return Cow::Borrowed(s),
+            None => id,
+        };
+        match id.strip_prefix(BODHI_UPDATE_PREFIX) {
+            Some(_) => Cow::Borrowed(id),
+            None => Cow::Owned(format!("{}{}", BODHI_UPDATE_PREFIX, id)),
+        }
+    }
+
+    pub(crate) fn get_update(updateid: &str) -> Result<BodhiUpdate> {
+        let updateid = canonicalize_update_id(updateid);
+        let url = format!("{}{}", BODHI_URL_PREFIX, updateid);
+        let update_data = crate::utils::download_url_to_tmpfile(&url, false)?;
+        let resp: BodhiUpdateResponse = serde_json::from_reader(update_data)?;
+        Ok(resp.update)
+    }
+
+    pub(crate) fn get_rpm_urls_from_update(updateid: &str, arch: &str) -> Result<Vec<String>> {
+        let update = bodhi::get_update(updateid)?;
+        update.builds.iter().try_fold(Vec::new(), |mut r, buildid| {
+            // For now hardcode skipping debuginfo because it's large and hopefully
+            // people aren't layering that.
+            let rpms = koji::get_rpm_urls_from_build(&buildid.nvr, arch, true)?;
+            r.extend(rpms);
+            Ok(r)
+        })
+    }
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        #[test]
+        fn test_canonicalize() {
+            let regid = "FEDORA-2020-053d8a2e94";
+            let shortid = "2020-053d8a2e94";
+            let url = "https://bodhi.fedoraproject.org/updates/FEDORA-2020-053d8a2e94";
+            assert_eq!(canonicalize_update_id(regid), regid);
+            assert_eq!(canonicalize_update_id(url), regid);
+            assert_eq!(canonicalize_update_id(shortid), regid);
+        }
+    }
+}
+
+mod koji {
+    use super::*;
+    use std::collections::BTreeMap;
+
+    #[derive(Default, Deserialize)]
+    #[serde(rename_all = "kebab-case")]
+    #[allow(dead_code)]
+    pub(crate) struct KojiBuildInfo {
+        nvr: String,
+        id: u64,
+        kojipkgs_url_prefix: String,
+        rpms: BTreeMap<String, Vec<String>>,
+    }
+
+    impl KojiBuildInfo {
+        fn rpmurl(&self, arch: &str, rpm: &str) -> String {
+            format!("{}/{}/{}", self.kojipkgs_url_prefix, arch, rpm)
+        }
+    }
+
+    pub(crate) fn get_buildid_from_url(url: &str) -> Result<&str> {
+        let id = url.rsplit('?').next().expect("split");
+        if !id.starts_with("buildID=") {
```

Comment:
Another really cool future addition is support for taskID URLs. Then it'd be super easy to test out scratch builds too.